cluster-autoscaler/cloudprovider/mcm/fakeclient/client.go

```diff
@@ -0,0 +1,592 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fakeclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	fakeuntyped "github.com/gardener/machine-controller-manager/pkg/client/clientset/versioned/fake"
+	apipolicyv1beta1 "k8s.io/api/policy/v1beta1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/discovery"
+	fakediscovery "k8s.io/client-go/discovery/fake"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	policyv1beta1 "k8s.io/client-go/kubernetes/typed/policy/v1beta1"
+	fakepolicyv1beta1 "k8s.io/client-go/kubernetes/typed/policy/v1beta1/fake"
+	k8stesting "k8s.io/client-go/testing"
+)
+
+// FakeObjectTracker implements both k8stesting.ObjectTracker as well as watch.Interface.
+type FakeObjectTracker struct {
+	*watch.FakeWatcher
+	delegatee    k8stesting.ObjectTracker
+	watchers     []*watcher
+	trackerMutex sync.Mutex
+	fakingOptions
+}
+
+// Add receives an add event with the object
+func (t *FakeObjectTracker) Add(obj runtime.Object) error {
+	if t.fakingOptions.failAll != nil {
+		err := t.fakingOptions.failAll.RunFakeInvocations()
+		if err != nil {
+			return err
+		}
+	}
+
+	return t.delegatee.Add(obj)
+}
+
+// Get receives a get event with the object
+func (t *FakeObjectTracker) Get(gvr schema.GroupVersionResource, ns, name string) (runtime.Object, error) {
+	var err error
+	if t.fakingOptions.failAll != nil {
+		err = t.fakingOptions.failAll.RunFakeInvocations()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if t.fakingOptions.failAt != nil {
+		if gvr.Resource == "nodes" {
+			err = t.fakingOptions.failAt.Node.Get.RunFakeInvocations()
+		} else if gvr.Resource == "machines" {
+			err = t.fakingOptions.failAt.Machine.Get.RunFakeInvocations()
+		} else if gvr.Resource == "machinesets" {
+			err = t.fakingOptions.failAt.MachineSet.Get.RunFakeInvocations()
+		} else if gvr.Resource == "machinedeployments" {
+			err = t.fakingOptions.failAt.MachineDeployment.Get.RunFakeInvocations()
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return t.delegatee.Get(gvr, ns, name)
+}
+
+// Create receives an create event with the object
+func (t *FakeObjectTracker) Create(gvr schema.GroupVersionResource, obj runtime.Object, ns string) error {
+	if t.fakingOptions.failAll != nil {
+		err := t.fakingOptions.failAll.RunFakeInvocations()
+		if err != nil {
+			return err
+		}
+	}
+
+	err := t.delegatee.Create(gvr, obj, ns)
+	if err != nil {
+		return err
+	}
+
+	if t.FakeWatcher == nil {
+		return errors.New("Error sending event on a tracker with no watch support")
```

Comment:
watch for goland warnings so that these small things can be taken care before filing the PR.
errors.New recommends to use small case starting letter.